### PR TITLE
[fix] allow DevCycle Android Client SDK to be consumed by apps that use minification and shrinking

### PR DESCRIPTION
**Pull request by kaushalkapasi**
- replace mixed usage of Jackson and GSON with only GSON
  - update API models to use GSON
- added tests to DVC Client Tests to make sure the SDK defaults variables when a null config is received

### CODE REVIEW

File: android-client-sdk/consumer-rules.pro [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Add build specific ProGuard rules here.
+# These rules will be pulled when a consumer installs the DevCycle Android SDK.
+# You can control the set of applied configuration files using the
+# proguardFiles setting in build.gradle.
+#
+# For more details, see
+#   http://developer.android.com/guide/developing/tools/proguard.html
+
+-keepclassmembers enum com.devcycle.sdk.** { *; }
```
Comment (VicV) [inline android-client-sdk/consumer-rules.pro:9]: why this

Comment (kaushalkapasi) [reply android-client-sdk/consumer-rules.pro]: good point, going to narrow it down to just keep the models instead of the entire SDK

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCApiClient.kt [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
+import retrofit2.converter.gson.GsonConverterFactory
 
 internal class DVCApiClient {
     private val okBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
-    private val objectMapper = jacksonObjectMapper()
     private val adapterBuilder: Retrofit.Builder = Retrofit.Builder()
-        .addConverterFactory(JacksonConverterFactory.create(objectMapper))
 
-    fun initialize(baseUrl: String): DVCApi {
+    fun initialize(baseUrl: String, gsonMapper: Gson): DVCApi {
```
Comment (jonathannorris) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCApiClient.kt:12]: is this a required field? why is this needed on the `initialize` function?

Also, would this require this change to be a major version?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Event.kt [hunk 3]
```diff
@@ -1,20 +1,26 @@
...
 
-import com.fasterxml.jackson.annotation.JsonInclude
 import com.fasterxml.jackson.annotation.JsonProperty
+import com.google.gson.annotations.SerializedName
 import java.math.BigDecimal
 import java.util.*
 
-@JsonInclude(JsonInclude.Include.NON_NULL)
 internal data class Event private constructor(
+    @SerializedName("type")
```
Comment (jonathannorris) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Event.kt:9]: How can we test that all these JSON serialization changes are working before release?

Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Event.kt:9]: both unit test & manual testing have covered them.. both in minified mode & normal, uncompressed mode! we're seeing the inbound events as expected and all other API calls / responses being parsed correctly

**Review by jonathannorris — APPROVED** [review]
Looks good to me, let's get someone who has worked more on this SDK to approve as well.

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt [hunk 4]
```diff
@@ -2,54 +2,64 @@
...
+    @SerializedName("deviceModel")
     val deviceModel: String = Build.MODEL,
     @Schema(description = "DevCycle SDK type")
+    @SerializedName("sdkType")
     val sdkType: String = "mobile",
     @Schema(description = "DevCycle SDK Version")
+    @SerializedName("sdkVersion")
     val sdkVersion: String = BuildConfig.VERSION_NAME,
     @Schema(description = "Date the user was last seen, Unix epoch timestamp format")
+    @SerializedName("lastSeenDate")
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt:62]: does gson not have a way to infer this automatically? seems weird that jackson can do this but gson can't

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt [hunk 5]
```diff
@@ -66,17 +55,26 @@ class Variable<T> internal constructor(
...
+    @SerializedName("isDefaulted")
     var isDefaulted: Boolean? = null
 ) : PropertyChangeListener {
 
     /**
      * Variable type
      */
     enum class TypeEnum(@get:JsonValue val value: String) {
-        STRING("String"), BOOLEAN("Boolean"), NUMBER("Number"), JSON("JSON");
+        @SerializedName("0")
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt:70]: whats going on here?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt (hunk 4 above)
Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt:62]: it isn't required, but better to apply it consistently across the board as I found it

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt (hunk 5 above)
Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt:70]: lets GSON understand the parsing of the `enum`

Comment (ajwootto) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt:70]: what's the significance of the serialized name being a specific number? How does that correspond to a type field with value `Boolean` being deserialized to `TypeEnum.Boolean`?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt [hunk 6]
```diff
@@ -16,12 +17,36 @@ import timber.log.Timber
...
-    private val api: DVCApi = DVCApiClient().initialize(apiBaseUrl)
     private val eventApi: DVCEventsApi = DVCEventsApiClient().initialize(envKey, eventsBaseUrl)
     private val edgeDBApi: DVCEdgeDBApi = DVCEdgeDBApiClient().initialize(envKey, apiBaseUrl)
     private val objectMapper = jacksonObjectMapper()
+    private val gsonMapper = getGsonMapper()
+    private val api: DVCApi = DVCApiClient().initialize(apiBaseUrl, gsonMapper)
     private val configMutex = Mutex()
 
+    private fun getGsonMapper(): Gson {
+        return GsonBuilder().registerTypeAdapter(Variable.TypeEnum::class.java, object : TypeAdapter<Variable.TypeEnum>() {
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:28]: is all this necessary to make the enums work?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt (hunk 5 above)
Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt:70]: Oh, actually.. reading the latest docs.. this isn't required it seems 👀

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt (hunk 6 above)
Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:28]: let me double check, we need this type adapter change for sure.. are we missing any other TypeEnums? 🤔 

- but I don't believe we need the `@SerializedNames("0")`, going to verify that

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt (hunk 5 above)
Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/Variable.kt]: yeah, works without this.. going to remove this annotation :)

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt (hunk 6 above)
Comment (ajwootto) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:28]: any idea why we need it? does gson not auto-infer stuff from the enum types?

Comment (kaushalkapasi) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:28]: yeah, it doesn't auto-infer from enums 😢 our unit tests caught this

Comment (VicV) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:28]: Specifically enums are an issue yes.
